### ISSUE TITLE
MSIS: previous day's F10.7

### ISSUE THREAD

**Issue by bharding512** [timduly4/pyglow#34]
I may be misunderstanding this, but I think the F10.7 input for MSIS should be the *previous* day's daily F10.7, not today's. It seems like pyglow is using today's.

From: http://download.hao.ucar.edu/pub/stans/msis/nrlmsise00.f
```
C     INPUT VARIABLES:
C        IYD - YEAR AND DAY AS YYDDD (day of year from 1 to 365 (or 366))
C              (Year ignored in current model)
C        SEC - UT(SEC)
C        ALT - ALTITUDE(KM)
C        GLAT - GEODETIC LATITUDE(DEG)
C        GLONG - GEODETIC LONGITUDE(DEG)
C        STL - LOCAL APPARENT SOLAR TIME(HRS; see Note below)
C        F107A - 81 day AVERAGE OF F10.7 FLUX (centered on day DDD)
C        F107 - DAILY F10.7 FLUX FOR PREVIOUS DAY
```

I'm not sure what IRI expects.

**Comment by timduly4** [comment]
Thanks for the catch and fix, @bharding512 .  Do I have to do my dissertation over again!? 😱 

IRI can accept a user input for F10.7 as a `jf` switch, but we don't use it:

```
C    i       .true.                  .false.          standard version
C    -----------------------------------------------------------------
...
C   25    F107D from APF107.DAT  F107D user input (oarr(41))         t
...
```
Where we use the `jf` switches: https://github.com/timduly4/pyglow/blob/master/pyglow/pyglow.py#L203

**Comment by bharding512** [comment]
Does this mean IRI has its own internal database of F10.7? Is that a concern? 

I'm thinking of a case where the user sets `user_ind=True`, then runs IRI and MSIS. It's possible the geophysical conditions are very different. 

Perhaps we should flip `jf[25]` and input F10.7 in that case.

**Comment by timduly4** [comment]
Good idea.  We first need to assign `self.user_ind = user_ind` in `Point.__init__()`.  Then, within IRI we can check for `self.user_ind` and if so, change the logic on `jf[35]`.  I can do a PR this weekend, but if you want to beat me to it, feel free 😁

**Comment by timduly4** [comment]
Hey @bharding512 , I've implemented the user indices for IRI: https://github.com/timduly4/pyglow/pull/40

However, there seems to be no effect on having your own F10.7 and F10.7A indices when calling IRI.  Within the `iri-user-ind` branch, check out `./tests/test_user_indices.py`:

```
Testing user defined indices in IRI:
------- ---- ------- ------- -- ----
Using default F10.7:
f10.7 = 125.60
f10.7a = 160.55
ne = 361218.21

Using user defined F10.7:
f10.7 = 200.00
f10.7a = 200.00
ne = 361218.21
```

Notice how the densities are the same.  Would you mind checking my work and seeing if anything is awry? (Specifically, the `jf` and `oarr` indices within `irisub.for`)

Note: You can easily install this version of pyglow with:
```
~ $ cd ~/pyglow
~/pyglow $ git pull
~/pyglow $ git checkout iri-user-ind
~/pyglow $ python ./setup.py install --user --prefix=
```
(This prevents downloading and recompiling all of the climatological models)

You can also easily see what I changed at: https://github.com/timduly4/pyglow/pull/40/files

Thanks,
Tim

**Comment by bharding512** [comment]
That's strange. I'll try to take a look, but with work and an upcoming vacation, it probably won't be for a couple weeks.

**Comment by timduly4** [comment]
OK, thanks.  @butala may have some insight, too.

**Comment by butala** [comment]
I think this is a bug in IRI. The critical line is in irisub.for:

        if(sam_date.and..not.rzino.and..not.rzin.
     &                   and..not.igin.and..not.igino) goto 2910

If this call is the same date is last time (year and DOY) and rz12 not given by the user (not sure what this is) and ig12 not given by the user (not sure what this is) then skip past using the user specified f107. This logic is also influenced by a global count of how many time the irisub subroutine has been called (see variable icalls).

I could hack something together but could never be confident that there would be no unintended consequences. IRI, like much fortran code in my opinion, is a spaghettified maze of awful.

The fix would seem to be to change the line around 437 of irisub.for to 

       if(icalls.ge.0) then

Again, I am not absolutely sure if this is correct in all possible cases.

**Comment by butala** [comment]
Hmm. The plot thickens. Take a look at iritest.for in iri2012. There is certain global state that is to be set before calls to iri:

		icalls=0
		nmono=-1
		iyearo=-1
		idaynro=-1
		rzino=-1
		igino=-1
		ut0=-1

The common block access and setting these state values was commented out in IRI2016. I think commenting this out is incorrect!

**Comment by butala** [comment]
Quick note. Did my best to reconcile everything. I sent an email to Prof. Bilitza regrading the code smell I am getting from IRI2016. A little hacking has brought everything into line as best as I can tell. Here is the current output of `./tests/test_user_indices.py`:

```
pyglow version = 1.1
Testing user defined indices in IRI:
------- ---- ------- ------- -- ----
 f107d   125.599998     f107y   125.800003
 f10781   160.500000     f107365   173.800003
Using default F10.7:
f10.7 = 125.60
f10.7a = 160.55
ne = 361218.21

 f107d   200.000000     f107y   200.000000
 f10781   200.000000     f107365   200.000000
Using user defined F10.7:
f10.7 = 200.00
f10.7a = 200.00
ne = 363029.59

 f107d   125.599998     f107y   125.599998
 f10781   160.550003     f107365   160.550003
Using user defined F10.7:
f10.7 = 125.60
f10.7a = 160.55
ne = 360263.09
```

Ideally the first and third test cases would be identical. However, the user cannot specify the `f107y` or `f107365` , i.e., they get set to the values of `f107d` and `f10781` (the user specified indices). I hardcoded all values to test that the IRI code would reproduce the same result in cases 1 and 3 and this was confirmed.

**Comment by butala** [comment]
I received a response from Prof. Bilitiza. See below. I need to go back and undo the change and we should update to the newer version of IRI that includes the fix.

Dear Mark,

Thank you for pointing out this problem. I have made changes to the program so it does not skip over the F10.7 input part. This was an oversight when introducing this new option of F10.7 user input. You can download the newest version from irimodel.org

icalls=0 is still being set in the DATA statement on Line 408. And the assignments are being made on lines 458-459 (see below). So you do not need to un-comment the COMMON/CONST2.

The changes are as follows:

Line 389:
     LOGICAL  EXT,SCHALT,TECON(2),sam_mon,sam_yea,sam_ut,sam_date,
    &  F1REG,FOF2IN,HMF2IN,URSIF2,LAYVER,RBTT,DREG,rzino,FOF1IN,
    &  HMF1IN,FOEIN,HMEIN,RZIN,sam_doy,F1_OCPRO,F1_L_COND,NODEN,
    &  NOTEM,NOION,TENEOP,OLD79,JF(50),URSIFO,igin,igino,mess,
    &  dnight,enight,fnight,fstorm_on,estorm_on,B0IN,B1IN,
    &  fof2ino,hmf2ino,f107in,f107ino,f107_81in,f107_81ino

Line 458-459:
       if(icalls.lt.1) then
               ARGMAX=88.0
               pi=ATAN(1.0)*4.
               UMR=pi/180.
               humr=pi/12.
               dumr=pi/182.5
               ALOG2=ALOG(2.)
               ALG10=ALOG(10.)
               ALG100=ALOG(100.)
               montho=-1
                       nmono=-1
                       iyearo=-1
                       idaynro=-1
                       rzino=.true.
                       igino=.true.
                       f107ino=.true.
                       f107_81ino=.true.
                       fof2ino=.true.
                       hmf2ino=.true.
                       ut0=-1
                       ursifo=.true.


Lines 955-965:
5592    continue
       sam_mon=(month.eq.montho)
       sam_yea=(iyear.eq.iyearo)
       sam_doy=(daynr.eq.idaynro)
       sam_date=(sam_yea.and.sam_doy)
       sam_ut=(hourut.eq.ut0)

       if(sam_date.and..not.rzin.and..not.rzino
    &          .and..not.igin.and..not.igino
    &      .and..not.f107in.and..not.f107ino
    &          .and..not.f107_81in.and..not.f107_81ino) goto 2910

Lines 1364-1375:
       nmono=nmonth
       MONTHO=MONTH
       iyearo=iyear
       idaynro=daynr
       rzino=rzin
       igino=igin
       f107ino=f107in
       f107_81ino=f107_81in
       ut0=hourut
       foF2ino=foF2in
       hmF2ino=hmF2in

**Comment by timduly4** [comment]
@butala , we download IRI 2016 "on the fly" each time we install:

https://github.com/timduly4/pyglow/blob/master/pyglow/models/get_models.py#L149

So, no changes need to be made to the model code itself, right?

I believe we need to remove the patch that you created here: https://github.com/timduly4/pyglow/blob/master/pyglow/models/f2py/iri16/irisub.patch

Am I on the right track?

**Comment by butala** [comment]
@timduly4 --- yes, you have it. What you suggest should clear everything up: remove the ugly patch I added and run the test to make sure everything is all clear (my scan of the IRI modifications convinces me the issue is resolved, but better to be safe).

**Comment by timduly4** [comment]
@butala great, thanks.  I've started a PR at https://github.com/timduly4/pyglow/pull/45 and will be sure to test before merging into `master`.
